docs/DevGuide/Databox.md

```diff
@@ -0,0 +1,367 @@
+For a detailed description of SpECTRE's DataBox, see the
+\ref DataBoxGroup "DataBox documentation".
+
+## SimpleTags {#documentation_for_simple_tags}
+Just as we needed to modify our tags to make them compatible with
+`TaggedTuple`, we need to again modify them for use with DataBox. Our ordinary
+tags become SpECTRE's SimpleTags:
+
+\snippet Test_DataBoxDocumentation.cpp proper_databox_tags
+
+As seen above, SimpleTags have a `type` and a `name` in their struct.
+When creating tags for use with a DataBox, we must make sure to derive off
+of one of the existing DataBox tag types such as `db::SimpleTag`. We now
+create our first DataBox using these `SimpleTags`:
+
+\snippet Test_DataBoxDocumentation.cpp refined_databox
+
+We can get our quantities out of the DataBox by using `db::get`:
+
+\snippet Test_DataBoxDocumentation.cpp refined_databox_get
+
+So far, the usage of DataBox has been similar to the usage of TaggedTuple. To
+address the desire to combine the functionality of tags with the modularity of
+functions, DataBox provides ComputeTags.
+
+## ComputeTags {#documentation_for_compute_tags}
+ComputeTags are used to tag functions that are used in conjunction with a
+DataBox to produce a new quantity. ComputeTags look like:
+
+\snippet Test_DataBoxDocumentation.cpp compute_tags
+
+ComputeTags derive off of `db::ComputeTag`, and it is convenient to have them
+additionally derive off of an existing SimpleTag (in this case `Mass`) so that
+the quantity `MassCompute` can be obtained through the SimpleTag `Mass`.
+We use the naming convention `TagNameCompute` so that `TagNameCompute` and
+`TagName` appear next to each other in documentation that lists tags in
+alphabetical order.
+
+We have also added the type alias `argument_tags`, which is necessary in order
+to refer to the correct tagged quantities in the DataBox.
+
+\note
+The `tmpl::list` used in the type alias is a contiguous container only
+holding types. That is, there is no variable runtime data associated with it
+like there is for `std::tuple`, which is a container associating types with
+values. They are useful in situations when one is working with multiple tags at
+once.
+
+This compile time resolution of *metavariables* is a pattern that is used again
+and again in SpECTRE. Let us see how we can compute our beloved quantity
+of mass times acceleration:
+
+\snippet Test_DataBoxDocumentation.cpp compute_tags_force_compute
+
+And that's it! `db::get` utilizes the `argument_tags` specified in
+`ForceCompute` to determine which items to get out of the `refined_databox`.
+With the corresponding quantities in hand, `db::get` passes them as arguments
+to the `function` specified in `ForceCompute`. This is why every `ComputeTag`
+must have an `argument_tags` as well as a `function` specified; this is the
+contract with DataBox they must satisfy in order to enjoy the full benefits of
+DataBox's generality.
+
+## Mutating DataBox items {#documentation_for_mutate_tags}
```

Comment:
Somewhere in this section it would be useful to point out that any time any quantity in the DataBox is altered, it must be done so with a mutate call. In particular, it is possible to store pointers or references to the data stored in the DataBox, but using them to edit data in the DataBox would entirely undermine the compute tags architecture. Instead, the DataBox itself must perform all changes via mutate functions, so any compute item functions can be called on updated quantities if necessary.